File: merge-k-sorted-lists.cpp
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0) return nullptr;
		if (lists.size() == 1) return lists[0];

		int mid = lists.size() / 2;
		vector<ListNode*> lt(lists.begin(), lists.begin()+mid);
		vector<ListNode*> rt(lists.begin()+mid, lists.end());

		ListNode* left = mergeKLists(lt);
		ListNode* right = mergeKLists(rt);
		
		return merge(left, right);
    }

	ListNode* merge(ListNode* list1, ListNode* list2) {
		ListNode *dummy = new ListNode();
		ListNode *ptr = dummy;

		while ((list1 != nullptr) && (list2 != nullptr)) {
			if (list1->val <= list2->val) {
				ptr->next = list1;
				list1 = list1->next;
			} else {
				ptr->next = list2;
				list2 = list2->next;
			}
			ptr = ptr->next;
		}

        if (list1 != nullptr) ptr->next = list1;
        if (list2 != nullptr) ptr->next = list2;r

		return dummy->next;
	}
};

File: merge-k-sorted-lists.py
from __future__ import annotations
from typing import List, Any, Optional

class ListNode:
    def __init__(self, val: Any = 0, next: Optional[ListNode] = None):
        self.val = val
        self.next = next

class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if len(lists) == 0:
            return None
        if len(lists) == 1:
            return lists[0]

        mid = len(lists) // 2
        left = self.mergeKLists(lists[:mid])
        right = self.mergeKLists(lists[mid:])
        return self.merge(left, right)

    def merge(self, list1: ListNode, list2: ListNode) -> ListNode:
        dummy = ListNode()
        ptr = dummy

        while list1 and list2:
            if list1.val <= list2.val:
                ptr.next = list1
                list1 = list1.next
            else:
                ptr.next = list2
                list2 = list2.next
            ptr = ptr.next

        ptr.next = list1 or list2
        return dummy.next

